pkg/pipeline/encode/encode_prom_test.go

```diff
@@ -62,124 +66,241 @@ parameters:
-
-	eInfoBytes := generateCacheKey(&eInfo)
-	_, found := encodeProm.mCache.GetCacheEntry(string(eInfoBytes))
-	require.Equal(t, true, found)
-
 	// wait a couple seconds so that the entry will expire
 	time.Sleep(2 * time.Second)
 	encodeProm.mCache.CleanupExpiredEntries(encodeProm.expiryTime, encodeProm)
 	entriesMapLen = encodeProm.mCache.GetCacheLen()
 	require.Equal(t, 0, entriesMapLen)
 }
 
-func Test_EncodeAggregate(t *testing.T) {
+func Test_CustomMetric(t *testing.T) {
 	metrics := []config.GenericMap{{
-		"name":       "test_aggregate",
-		"operation":  "sum",
-		"record_key": "IP",
-		"by":         "[dstIP srcIP]",
-		"aggregate":  "20.0.0.2,10.0.0.1",
-		"value":      7.0,
-		"count":      1,
+		"srcIP":   "20.0.0.2",
+		"dstIP":   "10.0.0.1",
+		"flags":   "SYN",
+		"bytes":   7,
+		"packets": 1,
+		"latency": 0.1,
+	}, {
+		"srcIP":   "20.0.0.2",
+		"dstIP":   "10.0.0.1",
+		"flags":   "RST",
+		"bytes":   1,
+		"packets": 1,
+		"latency": 0.05,
+	}, {
+		"srcIP":   "10.0.0.1",
+		"dstIP":   "30.0.0.3",
+		"flags":   "SYN",
+		"bytes":   12,
+		"packets": 2,
+		"latency": 0.2,
 	}}
 
-	newEncode := &EncodeProm{
-		port:   ":0000",
-		prefix: "test_",
-		metrics: map[string]metricInfo{
-			"gauge": {
-				input: "value",
-				filter: keyValuePair{
-					key:   "name",
-					value: "test_aggregate",
-				},
-				labelNames: []string{"by", "aggregate"},
-			},
-		},
-		mCache: utils.NewTimedCache(),
+	params := api.PromEncode{
+		Port:       9090,
+		Prefix:     "test_",
+		ExpiryTime: 60,
+		Metrics: []api.PromMetricsItem{{
+			Name:     "bytes_total",
+			Type:     "counter",
+			ValueKey: "bytes",
+			Labels:   []string{"srcIP", "dstIP"},
+		}, {
+			Name:     "packets_total",
+			Type:     "counter",
+			ValueKey: "packets",
+			Labels:   []string{"srcIP", "dstIP"},
+		}, {
+			Name:     "latency_seconds",
+			Type:     "histogram",
+			ValueKey: "latency",
+			Labels:   []string{"srcIP", "dstIP"},
+			Buckets:  []float64{},
+		}},
 	}
 
-	newEncode.Encode(metrics)
+	encodeProm, cleanup, err := initPromWithServer(&params)
+	require.NoError(t, err)
+	defer cleanup()
+
+	encodeProm.Encode(metrics)
+	time.Sleep(100 * time.Millisecond)
 
-	gEntryInfo1 := config.GenericMap{
-		"Name": "test_gauge",
-		"Labels": map[string]string{
-			"by":        "[dstIP srcIP]",
-			"aggregate": "20.0.0.2,10.0.0.1",
-		},
-		"value": float64(7),
+	exposed := test.ReadExposedMetrics(t)
+
+	require.Contains(t, exposed, `test_bytes_total{dstIP="10.0.0.1",srcIP="20.0.0.2"} 8`)
+	require.Contains(t, exposed, `test_bytes_total{dstIP="30.0.0.3",srcIP="10.0.0.1"} 12`)
+	require.Contains(t, exposed, `test_packets_total{dstIP="10.0.0.1",srcIP="20.0.0.2"} 2`)
+	require.Contains(t, exposed, `test_packets_total{dstIP="30.0.0.3",srcIP="10.0.0.1"} 2`)
+	require.Contains(t, exposed, `test_latency_seconds_bucket{dstIP="10.0.0.1",srcIP="20.0.0.2",le="0.025"} 0`)
+	require.Contains(t, exposed, `test_latency_seconds_bucket{dstIP="10.0.0.1",srcIP="20.0.0.2",le="0.05"} 1`)
```

Comment:
Where are these buckets defined? Are they a default?

Reply:
yes, prom client defines default: `DefBuckets = []float64{.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10}`

This hasn't changed from the previous implementation